File: entsoe-client/src/main/java/org/example/client/EntsoEClientFactory.java
package org.example.client;

import org.example.config.EntsoEClientConfiguration;
import org.example.http.DefaultHttpClient;
import org.example.http.HttpClient;

public class EntsoEClientFactory {

    public static EntsoEClient createClient(String apiToken) {
        if (apiToken == null || apiToken.trim().isEmpty()) {
            throw new IllegalArgumentException("API token cannot be null or empty");
        }

        EntsoEClientConfiguration config = EntsoEClientConfiguration.defaultConfig(apiToken);
        return createClient(config);
    }

    public static EntsoEClient createClient(EntsoEClientConfiguration config) {
        if (config == null) {
            throw new IllegalArgumentException("Configuration cannot be null");
        }

        // Validate configuration properties
        if (config.getBaseUrl() == null || config.getBaseUrl().trim().isEmpty()) {
            throw new IllegalArgumentException("Base URL cannot be null or empty");
        }

        HttpClient httpClient = new DefaultHttpClient(config);
        return new DefaultEntsoEClient(httpClient, config.getBaseUrl());
    }

    public static EntsoEClient createClient(EntsoEClientConfiguration config, HttpClient customHttpClient) {
        if (config == null) {
            throw new IllegalArgumentException("Configuration cannot be null");
        }

        if (config.getBaseUrl() == null || config.getBaseUrl().trim().isEmpty()) {
            throw new IllegalArgumentException("Base URL cannot be null or empty");
        }

        return new DefaultEntsoEClient(customHttpClient, config.getBaseUrl());
    }
}

File: entsoe_client/src/entsoe_client/client/entsoe_client_factory.py
from entsoe_client.client.default_entsoe_client import DefaultEntsoEClient
from entsoe_client.client.entsoe_client import EntsoEClient
from entsoe_client.config.settings import EntsoEClientConfig
from entsoe_client.container import Container
from entsoe_client.exceptions.entsoe_client_factory_error import (
    EntsoEClientFactoryError,
)


class EntsoEClientFactory:
    """Factory for creating EntsoE clients."""

    @staticmethod
    def create_client(api_token: str) -> EntsoEClient:
        """
        Create an EntsoE client with the provided API token.

        Args:
            api_token: ENTSO-E API token for authentication

        Returns:
            Configured EntsoE client instance

        Raises:
            ValueError: If API token is None or empty
        """
        if not api_token or not api_token.strip():
            raise EntsoEClientFactoryError.api_token_empty()

        config = EntsoEClientConfig(api_token=api_token)

        # Create container with configuration
        container = Container()
        container.config.override(config)

        # Create HTTP client from container
        http_client = container.http_client()

        return DefaultEntsoEClient(http_client, str(config.base_url))
